### ISSUE TITLE
SAM3 concept/semantic masks are thresholded on raw logits without sigmoid (logits > 0.5 instead of sigmoid(logits) > 0.5)

### ISSUE THREAD

**Issue by JunseongAHN** [ultralytics/ultralytics#25378]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

# SAM3 concept/semantic masks are thresholded on raw logits without `sigmoid` (`logits > 0.5` instead of `sigmoid(logits) > 0.5`)

**Version:** `ultralytics` 8.4.104 · Component: `models/sam` SAM3 semantic path (`SAM3SemanticPredictor`)

### Bug

The SAM3 mask head emits **raw logits** (no activation). In the concept/semantic path they are binarized with a plain `> 0.5` **without `sigmoid`**, so the effective threshold is `sigmoid(logit) > 0.622` instead of the intended `0.5`. Masks come out **systematically eroded** (boundaries pulled inward, thin structures clipped).

The official HF `transformers` reference applies `sigmoid` before thresholding at the same default `mask_threshold=0.5`, and Ultralytics' own **interactive** SAM3 path thresholds correctly (`> mask_threshold`, default `0.0`) — so the two Ultralytics paths are inconsistent.

### Location — `ultralytics/models/sam/predict.py`

```python
# SAM3SemanticPredictor.postprocess (~L2328)
masks = F.interpolate(masks.float()[None], orig_img.shape[:2], mode="bilinear")[0] > 0.5
# SAM3SemanticPredictor.inference_features (~L2403)
pred_masks = F.interpolate(pred_masks.float()[None], src_shape[:2], mode="bilinear")[0] > 0.5
```

Official (`transformers/models/sam3/image_processing_sam3.py`): `pred_masks.sigmoid()` then `masks > mask_threshold` (default `0.5`).

### Fix

Add `.sigmoid()` before `> 0.5` in both locations (or threshold raw logits at `> 0.0`):

```diff
- ...mode="bilinear")[0] > 0.5
+ ...mode="bilinear")[0].sigmoid() > 0.5
```

### Environment
  - Runtime: Docker container
  - OS: Ubuntu 22.04.3 LTS (Jammy), x86_64
  - Kernel: Linux 4.15.0-213-generic
  - Python: 3.10.12
    - Executable: `/usr/bin/python3.10`
    - Compiler: GCC 11.4.0
  - pip: 26.1.2
    - Location: `/usr/local/lib/python3.10/dist-packages/pip`
  - Installation method: pip
  - Ultralytics: 8.4.104
  - PyTorch: 2.5.1+cu121
  - torchvision: 0.20.1+cu121
  - NumPy: 2.2.6
  - OpenCV: 5.0.0.93
  - Pillow: 12.2.0
  - CUDA available: Yes
  - PyTorch CUDA build: 12.1
  - NVIDIA driver: 535.146.02
  - Maximum CUDA version reported by `nvidia-smi`: 12.2
  - cuDNN: 9.1.0
  - CUDA toolkit / `nvcc`: Not installed or not available in PATH
  - GPUs: 4 × NVIDIA GeForce RTX 3090, 24 GB each
  - CPU: AMD Ryzen Threadripper PRO 3995WX
    - 64 physical cores / 128 logical CPUs
  - RAM: 125.8 GB
  - Swap: None
  - SAM3 input size: 1008
  - SAM3 weight: `sam3.pt`

  `ultralytics.checks()` output:

  ```text
  Ultralytics 8.4.104 🚀 Python-3.10.12 torch-2.5.1+cu121
  CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
  Setup complete ✅ (128 CPUs, 125.8 GB RAM)
  ```

### Minimal Reproducible Example
Reproduction (no weights needed)
```python
import torch
logits = torch.linspace(-3, 3, steps=13)      # raw mask-head logits
ultra    = logits > 0.5                        # Ultralytics semantic path (bug)
official = logits.sigmoid() > 0.5              # official  (== logits > 0.0)
print("disagree:", (ultra != official).int().tolist())
# pixels with logit in (0, 0.5] -> official=1, ultralytics=0 -> eroded
```
output
```python
logit          : [-3.0, -2.5, -2.0, -1.5, -1.0, -0.5, 0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0]
sigmoid(logit) : [0.047, 0.076, 0.119, 0.182, 0.269, 0.378, 0.5, 0.622, 0.731, 0.818, 0.881, 0.924, 0.953]
ultralytics    : [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1]
official       : [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1]
disagree       : [0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0]
```

### Are you willing to submit a PR?

- [o] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @JunseongAHN, thank you for the detailed SAM3 report and the included minimal reproduction example 🚀. The comparison between the semantic and interactive paths, along with the environment details, is very helpful for investigation.

This is an automated response 🤖, and an Ultralytics engineer will also assist with reviewing the reported behavior soon. In the meantime, please confirm whether the issue is reproducible with the latest package and share any additional mask visualizations or logs if available.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and answers to common questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) to help us debug it. Your report already includes a focused MRE; a complete SAM3 reproduction with the relevant input and output masks would also be useful if available.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello. Thanks for debugging and reporting.

**Comment by UltralyticsAssistant** [comment]
A potential fix has now been merged in [PR #25386](https://github.com/ultralytics/ultralytics/pull/25386) by [@Y-T-G](https://github.com/Y-T-G).

The update changes SAM3 semantic mask postprocessing to use the model’s configurable `mask_threshold` instead of a hardcoded `0.5` threshold. This has been applied to both standard prediction and feature-based inference while preserving the existing mask resizing and bounding-box scaling behavior. As a result, mask sensitivity can now be adjusted through the model configuration, which may resolve issues involving mask boundaries or unexpected foreground pixel classification.

To test the changes before the next official PyPI release, install the latest code from the `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Please test this version and share whether the changes resolve the issue. Thank you 🙏 for reporting it, and any additional feedback is welcome if the problem persists.
